### ISSUE TITLE
define semantics for floating-point min and max collective functions

### ISSUE THREAD

**Issue by bashbaug** [KhronosGroup/OpenCL-Docs#833]
The semantics for the floating-point min and max collective functions (e.g. `work_group_scan_inclusive_min`) are currently undefined, especially for NaNs.

https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_C.html#work-group-functions

Should these functions behave as though the operator were the `fmin` of `fmax` built-in funcitons, which have defined NaN behavior?  Example for `fmin`: "Returns y if y < x, otherwise it returns x. If one argument is a NaN, fmin() returns the other argument. If both arguments are NaNs, fmin() returns a NaN"

Or should these functions behave as though the operator were the `min` or `max` built-in functions, which have no defined NaN behavior?  Example for `min`: "Returns y if y < x, otherwise it returns x. If x or y are infinite or NaN, the return values are undefined."

Or, should these functions behave some other way?

For reference, the CTS tests appear to be using `std::min` and `std::max`, though perhaps this is incorrect:
https://github.com/KhronosGroup/OpenCL-CTS/blob/main/test_conformance/workgroups/test_wg_scan_reduce.cpp#L75

Once we have settled on what should be happening in OpenCL C we should document the SPIR-V behavior also, either in the main SPIR-V spec (by updating the descriptions for **OpGroupFMin** and **OpGroupFMax**), or by updating the OpenCL SPIR-V environment spec.

**Comment by SunSerega** [comment]
If you decide on `fmin`/`fmax` - would probably also help to define identity value of `NaN`. Of course work group of size 0 is not a thing. But, for instance, currently on up-to-date NV drivers for my GPU:

- `work_group_reduce_min(+1.0d, +2.0d)` is 0;
- `work_group_reduce_max(-1.0d, -2.0d)` is 2.2250738585072E-308;

Obviously wrong, by also technically allowed. By defining identity=`NaN` - this can be strongly disallowed.

**Comment by bashbaug** [comment]
> If you decide on `fmin`/`fmax` - would probably also help to define identity value of `NaN`.

Yes, agreed, this is an interesting question.  Both the OpenCL C and SPIR-V specs currently define the identity value as +INF (for min) and -INF (for max), but with these identity values and `fmin`/`fmax` semantics `work_group_reduce_min(NaN, NaN, ...)` would return +INF, which may not be intended?

PS: AFAICT NVIDIA doesn't support the work-group collective functions ([CL_DEVICE_WORK_GROUP_COLLECTIVE_FUNCTIONS_SUPPORT](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#CL_DEVICE_WORK_GROUP_COLLECTIVE_FUNCTIONS_SUPPORT) returns false on my test system, also with up-to-date drivers), so I'm a little surprised the cases above are compiling.

**Comment by SunSerega** [comment]
> PS: AFAICT NVIDIA doesn't support the work-group collective functions

Hmm yeah, the device property also returns false for me, but it still lets me use it like this:
```
    kernel void k(global double* a) {
        size_t id = get_local_id(0);
        
        a[id] = work_group_reduce_min(a[id]);
    }
```
(not even pragma for cl_khr_fp64 required!)

But oh well, this same driver (516.94 for 750Ti) also reports being an OpenCL3.0 with OpenCL-C1.2, but lets me compile with `-cl-std=CL3.0`. And it also lets me use OpenCL-C3.0 features without passing any version argument.

Guess they are still only wetting their feet with the whole 3.0 thing...
